citadel/indico_citadel/backend.py

```diff
@@ -0,0 +1,179 @@
+                                .format(response.status_code, response.text))
+        response.close()
+
+    def upload_file(self, entry, entries, session):
+        with entry.attachment.file.open() as file:
+            response = session.put(
+                url_join(self.search_app, f'api/record/{entry.search_id}/files/{entry.attachment.file.filename}'),
+                data=file
+            )
+        if response.ok:
+            entry.attachment_file_id = entry.attachment.file.id
+            db.session.merge(entry)
+            db.session.commit()
+        else:
+            self.logger.error('Failed uploading attachment %d', entry.attachment.id)
+
+    def run_initial(self, events, total=None):
+        cte = Category.get_tree_cte(lambda cat: db.func.json_build_object('id', cat.id, 'title', cat.title))
+        self.categories = dict(db.session.execute(select([cte.c.id, cte.c.path])).fetchall())
+        super().run_initial(events, total)
+
+    def upload_records(self, records, from_queue):
+        session = requests.Session()
+        retry = Retry(
+            total=5,
+            backoff_factor=30,
+            status_forcelist=[500, 502, 503, 504],
+            allowed_methods=frozenset(['POST', 'PUT', 'DELETE'])
+        )
+        session.mount(self.search_app, HTTPAdapter(max_retries=retry))
+        session.headers = self.headers
+        uploader = parallelize(self.upload_record, entries=records)
+        uploader(session, from_queue)
+
+    def upload_files(self, files):
+        session = requests.Session()
+        retry = Retry(
+            total=5,
+            backoff_factor=30,
+            status_forcelist=[500, 502, 503, 504],
+            allowed_methods=frozenset(['PUT'])
+        )
+        session.mount(self.search_app, HTTPAdapter(max_retries=retry))
+        session.headers = self.headers
+        uploader = parallelize(self.upload_file, entries=files, batch_size=100)
+        uploader(session)
+
+
+class LiveSyncCitadelBackend(LiveSyncBackendBase):
+    """Citadel
+
+    This backend uploads data to Citadel.
+    """
+
+    uploader = LiveSyncCitadelUploader
+    unique = True
+
+    def run_export_files(self, batch=1000, force=False):
+        from indico_citadel.plugin import CitadelPlugin
+
+        attachments = (
+            CitadelSearchAppIdMap.query
+            .join(Attachment)
+            .join(AttachmentFile, Attachment.file_id == AttachmentFile.id)
+            .filter(AttachmentFile.size <= 10 * 1024 * 1024)
+            .filter(AttachmentFile.extension.in_(CitadelPlugin.settings.get('file_extensions')))
+        )
+        if not force:
+            attachments = attachments.filter(CitadelSearchAppIdMap.attachment_file_id.is_(None))
+        uploader = self.uploader(self)
+        attachments = verbose_iterator(attachments.yield_per(batch), attachments.count(),
+                                       attrgetter('id'), lambda obj: str_to_ascii(getattr(obj, 'title', '')),
```

Comment:
I changed this upstream since `str_to_ascii` didn't properly avoid breaking some lines due to control/tab chars. Use this instead:

```py
lambda obj: re.sub(r'\s+', ' ', strip_control_chars(obj.attachment.title))
```